Initial implementation of Runaway Takeoff Prevention (anti-taz)

etracer65 opened the pull request:
Addresses https://github.com/betaflight/betaflight/issues/4815

Detects runaway pidSum values and auto-disarms to prevent the "Tasmanian Devil" caused by incorrect props, wrong motor order/direction, incorrect flight controller orientation, etc.  After a successful takeoff and normal flight is detected the feature is disabled for the remainder of the battery.

For the initial implementation there are a lot of CLI parameters that adjust the trigger and deactivation behavior.  After testing I expect that some of the parameters will be removed and replaced with hard-coded optimum values.

The following parameters relate to the triggering phase:

**runaway_takeoff_prevention** - ON/OFF (default ON).

**runaway_takeoff_threshold** - The pidSum threshold limit on any axis required to trigger prevention (default 60 - equates to 60% in blackbox or 600 raw value).

**runaway_takeoff_activate_delay** - The amount of time in milliseconds that the pidSum must be above the threshold for triggering (default 75).

The remaining parameters are used to determine "successful flight" and deactivate the feature for the remainder of the battery.  The deactivation works by accumulating in-flight time if a number of conditions are met.  Once this accumulated time reaches the configured value, the prevention is deactivated.  This way crashes or gate clips won't trigger a disarm.  We're looking for the throttle level, activity on the sticks, and the pidSum staying low.

**runaway_takeoff_deactivate_throttle_percent** - The minimum throttle percentage required to accumulate in-flight time (default 25).

~~**runaway_takeoff_deactivate_pidlimit** - The pidSum on every axis must be below this limit (default 10).~~  **Parameter removed (hard-coded to 10).**

~~**runaway_takeoff_deactivate_stick_percent** - The minimum stick deflection percentage for any axis (default 15).~~ **Parameter removed (hard-coded to 15).**

**runaway_takeoff_deactivate_delay** - The accumulated time limit for deactivation in milliseconds (default 500 meaning 0.5 seconds).

For testing there is a new debug mode added: `set debug_mode = RUNAWAY_TAKEOFF`

debug[0] - boolean indicating whether the feature is active and in detection mode

debug[1] - boolean indicating whether the pidSum has exceeded the threshold and we are in a countdown for runaway_takeoff_activate_delay milliseconds.

debug[2] - boolean indicating that the criteria for deactivation has been met and we are accumulating in-flight time.

debug[3] - number indicating the in-flight milliseconds accumulated so far.  Once this reaches runaway_takeoff_deactivate_delay the feature will be disabled.

Some caveats/known issues:

1. ~~3D mode is not properly supported yet.  Triggering will work but deactivation will happen early because the throttle percentage is not taking into account the dual direction.  This will be fixed.~~   **3D support fixed.**

2. It's possible to trigger the disarm by crashing immediately after liftoff before the deactivation occurs.  You will be able to rearm.  Although it's generally resistant, severe bouncing on takeoff could also trigger the disarm.  Increasing the runaway_takeoff_activate_delay can help with this at the risk of making the triggering during a real event slower and possibly less effective.

3. Arming with no props on and then significantly moving the sticks could trigger a disarm.  This is because the pidSum values are growing to large values because the quad is not responding to the command inputs.  There's nothing wrong with this but it's a point to make clear to avoid confusion.

mikeller commented on src/main/fc/fc_core.c:
```diff
@@ -434,6 +496,66 @@ void processRx(timeUs_t currentTimeUs)
+                && (fabsf(axisPID_P[FD_ROLL] + axisPID_I[FD_ROLL] + axisPID_D[FD_ROLL]) < pidSumLimit)
+                && (fabsf(axisPID_P[FD_YAW] + axisPID_I[FD_YAW]) < pidSumLimit)) {
+                
+                inFlight = true;
```
`inStableFlight` is probably more accurate for this.

etracer65 replied:
Done.

mikeller commented on src/main/fc/fc_core.c:
```diff
@@ -434,6 +496,66 @@ void processRx(timeUs_t currentTimeUs)
+            DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 2, 0);
+            DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 3, runawayTakeoffAccumulatedUs / 1000); // Accumulated in-flight time in ms
+        } else {
+            DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 2, 1);  // deactivation countdown is active
```
Better use defines instead of `0`, `1` here. This will also eliminate the need for the comment.

The same goes for the debug index, with defines it will be immediately clear to the reader what the logged values are for.

etracer65 replied:
Done.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -113,6 +113,13 @@ PG_DECLARE_ARRAY(pidProfile_t, MAX_PROFILE_COUNT, pidProfiles);
+    uint8_t runaway_takeoff_deactivate_throttle; // minimum throttle percent required during deactivation phase
+    uint8_t runaway_takeoff_deactivate_pidlimit; // pidSum limit during deactivation phase
+    uint8_t runaway_takeoff_deactivate_stick;    // minimum stick deflection during deactivation phase
+    uint16_t runaway_takeoff_deactivate_delay;   // delay in ms for "in-flight" conditions before deactivation (successful flight)
```
Move this up to below `runaway_takeoff_activate_delay` to prevent padding.

etracer65 replied:
Done.

McGiverGim commented:
Maybe is better to let it OFF until tested and adjusted? What do you think?

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -607,6 +728,41 @@ static void subTaskPidController(timeUs_t currentTimeUs)
+        }
+        DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 0, 1);  // indicates whether prevention is active
+        if (runawayTakeoffTriggerUs == 0) {
+            DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 1, 0);
```
Use `DEBUG_SET(DEBUG_RUNAWAY_TAKEOFF, 1, runawayTakeoffTriggerUs == 0 ? 0 : 1);`

etracer65 replied:
Done.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -607,6 +728,41 @@ static void subTaskPidController(timeUs_t currentTimeUs)
+        && !runawayTakeoffCheckDisabled
+        && (!feature(FEATURE_MOTOR_STOP) || isAirmodeActive() || (calculateThrottleStatus() != THROTTLE_LOW))) {
+
+        float runawayTakeoffThreshold = pidConfig()->runaway_takeoff_threshold * 10.0f;
```
`const float runawayTakeoffThreshold = pidConfig()->runaway_takeoff_threshold * 10.0f;`

etracer65 replied:
Done.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
+
+// calculate the throttle percent based on stick position in the range from min_check to PWM_RANGE_MAX - integer math is good enough here.
+// TODO - make this work for 3D mode as well
+uint8_t calculateThrottlePercent(void) {
```
K&R style:
```
uint8_t calculateThrottlePercent(void)
{
```

etracer65 replied:
Done.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
+    } else {
+       result = ((rcData[THROTTLE] - rxConfig()->mincheck) * 100) / (PWM_RANGE_MAX - rxConfig()->mincheck);
+    }
+    return result;
```
```
const uint ret = constrain(100*(rcData[THROTTLE] - rxConfig()->mincheck) / (PWM_RANGE_MAX - rxConfig()->mincheck), 0, 100);
```

etracer65 replied:
This is a good simplification.  Originally did it literally because I knew I had to add 3D support later wanted it easy to understand.  Also because of upcoming 3D support I left the `ret` variable in place rather than just returning the calculation.  It will get more complex having to compare the 3D throttle deadbands, etc.

```
uint8_t calculateThrottlePercent(void)
{
    uint8_t ret;

    ret = constrain(((rcData[THROTTLE] - rxConfig()->mincheck) * 100) / (PWM_RANGE_MAX - rxConfig()->mincheck), 0, 100);
    return ret;
}
```

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
+    bool result = false;
+
+    for (int axis = FD_ROLL; axis <= FD_YAW; axis ++) {
+        if (axis == FD_YAW) {
```
`const int deadband = axis == FD_YAW ? rcControlsConfig()->yaw_deadband : rcControlsConfig()->deadband;`

etracer65 replied:
Done.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
+        } else {
+            deadband = rcControlsConfig()->deadband;
+        }
+        if ((rcData[axis] >= PWM_RANGE_MAX) || (rcData[axis] <= PWM_RANGE_MIN)) {
```
declare `stickPercent` here.

etracer65 replied:
@martinbudden Not sure I understand.  `stickPercent` is declared at the top of the function.  Do you mean declare it inline when assigned?  When there are multiple occurrences I generally declare at the top to avoid having them scattered through the code.

```
bool areSticksActive(uint8_t stickPercentLimit)
{
    uint8_t stickPercent;
    bool ret = false;

    for (int axis = FD_ROLL; axis <= FD_YAW; axis ++) {
        const uint8_t deadband = axis == FD_YAW ? rcControlsConfig()->yaw_deadband : rcControlsConfig()->deadband;
        if ((rcData[axis] >= PWM_RANGE_MAX) || (rcData[axis] <= PWM_RANGE_MIN)) {
            stickPercent = 100;
        } else {
            if (rcData[axis] > (rxConfig()->midrc + deadband)) {
                stickPercent = ((rcData[axis] - rxConfig()->midrc - deadband) * 100) / (PWM_RANGE_MAX - rxConfig()->midrc - deadband);
            } else if (rcData[axis] < (rxConfig()->midrc - deadband)) {
                stickPercent = ((rxConfig()->midrc - deadband - rcData[axis]) * 100) / (rxConfig()->midrc - deadband - PWM_RANGE_MIN);
            } else {
                stickPercent = 0;
            }
        }
        if (stickPercent >= stickPercentLimit) {
            ret = true;
        }
    }
    return ret;
}
```

mikeller replied:
Coding convention for Betaflight is to declare variables when they are first used, except for local scope static variables, which are declared at the beginning of the function. This makes code easier to maintain, and also somewhat easier to read.

etracer65 replied:
Ok.  I've always tended to do all my declarations at the top of the function but I guess that's kind of "old school" (like me!).  I don't have a problem changing.  In this case it's a little awkward with the scope of `stickPercent` is in 4 places so I rearranged a little.  It just feels weird to me to simple declarations (without assigning a value) inline.  To me they seem to get lost in the code.  How does this look instead?

```
// determine if the R/P/Y stick deflection exceeds the specified limit - integer math is good enough here.
bool areSticksActive(uint8_t stickPercentLimit)
{
    bool ret = false;

    for (int axis = FD_ROLL; axis <= FD_YAW; axis ++) {
        const uint8_t deadband = axis == FD_YAW ? rcControlsConfig()->yaw_deadband : rcControlsConfig()->deadband;
        uint8_t stickPercent = 0;
        if ((rcData[axis] >= PWM_RANGE_MAX) || (rcData[axis] <= PWM_RANGE_MIN)) {
            stickPercent = 100;
        } else {
            if (rcData[axis] > (rxConfig()->midrc + deadband)) {
                stickPercent = ((rcData[axis] - rxConfig()->midrc - deadband) * 100) / (PWM_RANGE_MAX - rxConfig()->midrc - deadband);
            } else if (rcData[axis] < (rxConfig()->midrc - deadband)) {
                stickPercent = ((rxConfig()->midrc - deadband - rcData[axis]) * 100) / (rxConfig()->midrc - deadband - PWM_RANGE_MIN);
            }
        }
        if (stickPercent >= stickPercentLimit) {
            ret = true;
        }
    }
    return ret;
}
```

mikeller replied:
Looks good to me. The assignment of a default value to `stickPercent` isn't required, but it will be optimised away by the compiler anyway.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
+uint8_t areSticksActive(uint8_t stickPercentLimit) {
+    uint8_t deadband;
+    uint8_t stickPercent;
+    bool result = false;
```
Use `ret` for variable that contains return value.

etracer65 replied:
Done.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -389,13 +402,61 @@ static bool canUpdateVTX(void)
 #endif
 
+// determine if the R/P/Y stick deflection exceeds the specified limit - integer math is good enough here.
+uint8_t areSticksActive(uint8_t stickPercentLimit) {
```
```
bool areSticksActive(uint8_t stickPercentLimit)
{
```

etracer65 replied:
Yeah, this was a mistake leftover from earlier revisions.  Originally I had a function that returns the stick percent for a supplied axis and the calling code would compare the limit.  Simplified to make one call and that checked all axis and returned boolean but forgot to update the return data type.  I'll fix.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -434,6 +495,66 @@ void processRx(timeUs_t currentTimeUs)
+                
+                inFlight = true;
+                if (runawayTakeoffDeactivateUs == 0) {
+                    runawayTakeoffDeactivateUs = currentTimeUs;
```
set `runawayTakeoffDeactivateUs = currentTimeUs + pidConfig()->runaway_takeoff_deactivate_delay * 1000;`
Then later your check can be simplified.

etracer65 replied:
See later comment, but I don't think this would work because of the needed accumulation logic.  However I think I can adopt this kind of logic for the detection phase inside of `static void subTaskPidController(timeUs_t currentTimeUs)`.  It's using a simple re-settable time limit.

martinbudden commented on src/main/fc/fc_core.c:
```diff
@@ -434,6 +495,66 @@ void processRx(timeUs_t currentTimeUs)
+
+        } else {
+            if (runawayTakeoffDeactivateUs != 0) {
+                runawayTakeoffAccumulatedUs += cmpTimeUs(currentTimeUs, runawayTakeoffDeactivateUs);
```
It seems unnecessary to accumulate time like this. Just set the time from when you started and compare with `currentTimeUs`

mikeller replied:
That would change the way this works though. The current implementation registers a successful flight and deactivates the check even if the pidsums exceed the limit or the sticks returned to their default positions at times. If this wasn't the case, then there would be a chance that the check never got deactivated during the flight.

martinbudden replied:
OK, I didn't look into the precise detail. It just seemed a bit strange.

etracer65 replied:
Correct.  The logic needs to accumulate multiple periods where the conditions are all true with possible gaps where they're not.  Originally the logic was a simple delay based on all the conditions being true for the full duration with a reset if not.  The necessity of adding stick activity made it difficult to find an appropriate delay time as all the conditions were only intermittently true.  Either we had to have too short a delay and risk false deactivation (the main problem), or deactivation would take too long (or not happen at all) because the conditions wouldn't be met contiguously for enough time - particularly throttle and stick deflection.  With the accumulation we gather time when stable flight appears to be underway until we have enough data to be confident.  Also the transitional events like throttle changes and stick deflection are what exacerbate the runaway so having multiple events while pidSum stays low really helps.

mikeller commented:
I noticed that tests are failing. Run `make test` to run them.

etracer65 commented:
@mikeller Not sure how to fix the failing tests.

```
$ make test
cd src/test && make test
compiling ../main/sensors/boardalignment.c 
compiling ../main/common/maths.c 
compiling unit/alignsensor_unittest.cc 
compiling ../../obj/test/gtest-all.o 
compiling ../../obj/test/gtest_main.o 
linking ../../obj/test/gtest_main.a 
ar: creating archive ../../obj/test/gtest_main.a
a - ../../obj/test/gtest-all.o
a - ../../obj/test/gtest_main.o
linking ../../obj/test/alignsensor_unittest/alignsensor_unittest 
Running main() from gtest_main.cc
[==========] Running 8 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 8 tests from AlignSensorTest
[ RUN      ] AlignSensorTest.ClockwiseZeroDegrees
[       OK ] AlignSensorTest.ClockwiseZeroDegrees (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseNinetyDegrees
[       OK ] AlignSensorTest.ClockwiseNinetyDegrees (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseOneEightyDegrees
[       OK ] AlignSensorTest.ClockwiseOneEightyDegrees (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseTwoSeventyDegrees
[       OK ] AlignSensorTest.ClockwiseTwoSeventyDegrees (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseZeroDegreesFlip
[       OK ] AlignSensorTest.ClockwiseZeroDegreesFlip (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseNinetyDegreesFlip
[       OK ] AlignSensorTest.ClockwiseNinetyDegreesFlip (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseOneEightyDegreesFlip
[       OK ] AlignSensorTest.ClockwiseOneEightyDegreesFlip (0 ms)
[ RUN      ] AlignSensorTest.ClockwiseTwoSeventyDegreesFlip
[       OK ] AlignSensorTest.ClockwiseTwoSeventyDegreesFlip (0 ms)
[----------] 8 tests from AlignSensorTest (0 ms total)

[----------] Global test environment tear-down
[==========] 8 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 8 tests.
running test_alignsensor_unittest: PASS
compiling ../main/fc/fc_core.c 
compiling ../main/fc/rc_controls.c 
compiling ../main/fc/rc_modes.c 
compiling ../main/fc/runtime_config.c 
compiling ../main/common/bitarray.c 
compiling unit/arming_prevention_unittest.cc 
linking ../../obj/test/arming_prevention_unittest/arming_prevention_unittest 
Undefined symbols for architecture x86_64:
  "_axisPID_D", referenced from:
      _processRx in fc_core.c.o
      _subTaskPidController in fc_core.c.o
  "_axisPID_I", referenced from:
      _processRx in fc_core.c.o
      _subTaskPidController in fc_core.c.o
  "_axisPID_P", referenced from:
      _processRx in fc_core.c.o
      _subTaskPidController in fc_core.c.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [../../obj/test/arming_prevention_unittest/arming_prevention_unittest] Error 1
make: *** [test] Error 2
```

The error is obviously referring to `axisPID_P[]` `axisPID_I[]` and `axisPID_D[]` which are defined in pid.h.

`extern float axisPID_P[3], axisPID_I[3], axisPID_D[3];`

pid.h is included in rc_core.c so not sure why they're undefined symbols.

ledvinap commented:
@etracer65 : `axisPID_[PID]` used to be only internal variable for blackbox logging ... Then it got used in mixer/servo, which is IMO bug (unnecessary breaking of encapsulation)

Stubbing variables in failing test should fix test error

etracer65 commented:
@McGiverGim This is a safety feature so having it enabled is preferred in my opinion.  Finding out you need to enable it after something bad has already happened defeats the purpose.

The worst negative effects expected are inadvertent disarming on rough takeoffs, and the logic not deactivating in a timely manner leading to a disarm on a subsequent crash after successful takeoff.  Part of the testing is to determine the optimal parameters if these are experienced and we need flight time to determine this.

Also testing needs to take place with `debug_mode = RUNAWAY_TAKEOFF` so we can examine the internal states of the logic in comparison to the events that triggered.

AndersHoglund commented:
Symbols are defined here, but this module is not part of the test build:
`src/main/flight/pid.c:FAST_RAM float axisPID_P[3], axisPID_I[3], axisPID_D[3];`

Should be stubbed in the test suite. Take a look at other stubs as examples.

etracer65 commented:
Ok, I stubbed in `float axisPID_P[3], axisPID_I[3], axisPID_D[3];` and that fixed `arming_prevention_unittest.cc` unit test.  But there's another failing test that's not related to my code (no mods to settings.h).  This unit test fails on the current master:

```
compiling unit/cli_unittest.cc 
unit/cli_unittest.cc:93:29: error: call to implicitly-deleted default
      constructor of 'const cliValueConfig_t'
    const clivalue_t cval = {
                            ^
../main/interface/settings.h:132:28: note: default constructor of '' is
      implicitly deleted because field 'lookup' has a deleted default
      constructor
    cliLookupTableConfig_t lookup;
                           ^
../main/interface/settings.h:124:30: note: default constructor of
      'cliLookupTableConfig_s' is implicitly deleted because field 'tableIndex'
      of const-qualified type 'const lookupTableIndex_e' would not be
      initialized
    const lookupTableIndex_e tableIndex;
                             ^
1 error generated.
make[1]: *** [../../obj/test/cli_unittest/cli_unittest.o] Error 1
make: *** [test] Error 2
```

AndersHoglund commented:
That is weird. `cli_unittes`t does not fail when running the tests on your branch. However, the `rc_controls_unittest` needed some stubbing too. Here are all I did to get all tests to run: 
```

$ git diff
diff --git a/src/test/unit/arming_prevention_unittest.cc b/src/test/unit/arming_prevention_unittest.cc
index 3ffd0ae..b4a1584 100644
--- a/src/test/unit/arming_prevention_unittest.cc
+++ b/src/test/unit/arming_prevention_unittest.cc
@@ -64,6 +64,7 @@ extern "C" {
     gpsSolutionData_t gpsSol;
     uint32_t targetPidLooptime;
     bool cmsInMenu = false;
+    float axisPID_P[3], axisPID_I[3], axisPID_D[3];
 }

 uint32_t simulationFeatureFlags = 0;
diff --git a/src/test/unit/rc_controls_unittest.cc b/src/test/unit/rc_controls_unittest.cc
index 3ade143..f46fdc6 100644
--- a/src/test/unit/rc_controls_unittest.cc
+++ b/src/test/unit/rc_controls_unittest.cc
@@ -49,6 +49,7 @@ extern "C" {
     #include "fc/rc_adjustments.h"

     #include "fc/rc_controls.h"
+    #include "fc/runtime_config.h"

     #include "scheduler/scheduler.h"
 }
@@ -56,6 +57,10 @@ extern "C" {
 #include "unittest_macros.h"
 #include "gtest/gtest.h"

+void unsetArmingDisabled(armingDisableFlags_e flag) {
+  UNUSED(flag);
+}
+
 class RcControlsModesTest : public ::testing::Test {
 protected:
     virtual void SetUp() {

```

etracer65 commented:
@AndersHoglund Yeah, I'm not sure what's going on.  I'm guessing a compiler version difference as I'm mainly working on a machine with a couple versions old OS X.  However I tested on a completely up-to-date OS X 10.13 machine and it didn't have the error with `cli_unittest.cc` from current master but did on my local branch.  So tested on yet another machine (this time OS X 10.10) and did not have the error with `cli_unittest.cc` at all.  So anyway I've made the necessary stubs to `arming_prevention_unittest.cc` and `rc_controls_unittest.cc` to get all tests to pass.

etracer65 commented:
Well even after adding:

```
@@ -64,6 +64,7 @@ extern "C" {
     gpsSolutionData_t gpsSol;
     uint32_t targetPidLooptime;
     bool cmsInMenu = false;
+    float axisPID_P[3], axisPID_I[3], axisPID_D[3];
 }
```

to `arming_prevention_unittest.cc` the unit test is still complaining about references to `axisPID_P` etc.  https://travis-ci.org/betaflight/betaflight/jobs/327537920

No errors in my local unit test run.  The other build errors are just space overflow for OMNIBUS, SPRACINGF3EVO, etc.

etracer65 commented:
@mikeller @martinbudden The requested changes have been committed.  Still one outstanding unit test error that doesn't want to be resolved and doesn't happen locally.

AndersHoglund commented:
Travis is weird ... at times. I pulled your latest updates and all unittests are OK here too. Will kick Travis in his b....s and see what happens.

AndersHoglund commented:
Travis is OK with the unittests now. But you have other issues with this PR, to big for some targets:
```
$ make AIORACERF3
Building AIORACERF3
/home/andho/projects/cleanflight/andwho/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: obj/main/betaflight_AIORACERF3.elf section `.data' will not fit in region `FLASH'
/home/andho/projects/cleanflight/andwho/betaflight/tools/gcc-arm-none-eabi-7-2017-q4-major/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: region `FLASH' overflowed by 658 bytes
collect2: error: ld returned 1 exit status
Makefile:285: recipe for target 'obj/main/betaflight_AIORACERF3.elf' failed

```

etracer65 commented:
@AndersHoglund Yes, we're aware of the overflow for some targets.  I think most of them were already overflowing before my pull request due to earlier PR's.

mikeller commented:
@etracer65: Looks good to me now - we'll just have to address the overflow on F3 targets.

@McGiverGim: Another reason to have this on even now that we know that the parameters are not fully tuned yet is that we **want** users to test this, and give us feedback that enables us to tune the parameters. The overarching convention for using latest on master is 'do it at your own risk', and we want to and need to use this resource to solicit as much and as widespread user feedback as we can - after all we don't have the money to pay for somebody to do QA, other than letting our users doing it for free.

etracer65 commented:
I'll start working on a wiki entry for this feature to document it's operation and the various CLI parameters.

It always frustrates me when a new feature is added and there's no clear source of documentation on how it works so I'll take the responsibility.

mikeller commented:
@etracer65: Now that we have restored the `docs/` folder you have a choice of either documenting it as a .md file (can be added to this pull request), or adding an entry to the wiki.

etracer65 commented:
I was able to fly the new code last night and everything worked perfectly!  In fact I'm amazed at how well it worked with basically no tuning.  The only adjustment made to the initial default settings was to increase the `runaway_takeoff_activate_delay` to 75ms as it turned out the the triggering did such a good job catching the spaz-out so quickly that there was head room to increase the bounce/bump resistance on takeoff.

Here's a short video with the flight controller orientation intentionally set to be 90 degrees offset.  Ends up being anticlimactic, but that's sort of the point!

https://youtu.be/FEczCjAWv_0

No matter how hard I tried I was unable to cause false triggering.  This included bouncing the quad 5 feet up in the air and bouncing off the battery repeatedly.  Also pitched forward with minimal throttle and "scooted" 20 feet across the floor bouncing about 10 times it it didn't even phase it.  Reviewing the blackbox logs showed the deactivation settings properly ignoring the quick pidSum spikes caused by the bounces.

I flew about 25 packs in racing conditions with lots of crashes (many recoveries with turtle mode!) and never experienced any negative effects.  In fact the best thing about it was that I got focused on the racing and completely forgot about the feature.  This is the best result as it was completely unobtrusive and only inserted itself when an actual runaway event occurred on takeoff.  This is key because we want to avoid pilots turning off the feature because of negative impact.  It should stay silently in the background until it needs to intervene to prevent a dangerous event.

Several other pilots were also running a build with the code and likewise they noticed no negative effects.  In fact we had an actual occurrence of one of the testing pilots mistakenly installing props swapped and the feature behaved exactly as it should - disarming immediately like in the video above.  Saved the pilot's quad from damage and protected the bystanders.  The characteristic sound and quick disarm made the other pilots immediately respond with "anti-taz!".

At this point I'm ready to remove a couple of CLI configurables as current values are optimal:

`runaway_takeoff_deactivate_pidlimit` - hard-code to 10
`runaway_takeoff_deactivate_stick_percent` - hard-code to 15

Eventually I think we'll also be able to remove `runaway_takeoff_threshold` and use a hard-coded value of 60, but I want to wait for more testing from the community.

braeden123 commented:
Wow, fantastic work, seriously! I'm glad to here that it works so well. Maybe ask some more pilots on the Slack to test out the feature just to be sure they can't come up with anything to trigger it.

etracer65 commented:
@braeden123 Once the feature is merged I'll be able to encourage people in the rcgroups thread to test as they'll have easier access to the hex files.  In past experience this will get the largest testing response.

etracer65 commented:
@mikeller We've previously discussed the unfortunate length of the CLI parameters and the desire to shorten them in a meningful way.  In addition to being nicer to type, there's the savings in flash space.

So my proposal is to move away from "runaway takeoff" and simply start calling it "anti-taz".  I originally wanted to call it this but restrained myself because I thought we'd want something less tongue-in-cheek.

The reality I'm finding is that everybody who knows about this upcoming feature is already calling it "anti-taz" anyway and I'm sure it will always be called this regardless of what we come up with.  Just like nobody ever refers to "flip over after crash" - they simply say "turtle mode!"

So I'm proposing something like:
`antitaz_mode = ON`
`antitaz_threshold = 60`

and so on.  Thoughts?

martinbudden commented:
@etracer65 , great stuff. Good to see all your hard work paid off.

`antitaz_` as a prefix seems fine to me.

fiveangle commented:
Great work and a huge jump in safety for all @etracer65 !

mikeller commented:
@etracer65: That's very promising indeed! Have you had time to look into covering stick activity detection when in 3D mode? I think if we don't do this now, we should probably change the code to not use the runaway takeoff detection to be not active when in 3d mode until it gets added, rather than have it work in unexpected ways when in 3d mode.

Re renaming to `antitaz_`, I am opposed to using non-descriptive names. the example of `moron_threshold` and the weekly user questions about its function can serve as an example of what this leads to.

mikeller commented:
As a more descriptive alternative, how about `freakout_stopper_`?

etracer65 commented:
> As a more descriptive alternative, how about `freakout_stopper_`?

That's actually longer than the current `runaway_takeoff`.  Also I don't think "freakout" clearly identifies the event at takeoff.  Could just as easily describe a de-sync behavior to a new user.

In support of `antitaz_` - everybody I've talked to or has read a post where the term "anti-taz" was used knew immediately that I was talking about the "Tasmanian Devil" on takeoff.  On the other hand I've had to continuously explain "runaway takeoff".  Also the intent is that these aren't parameters that we would encourage users to change.  The ones that do should have a specific reason and a good understanding of what they're changing.

Lastly having "ANTI-TAZ" show up in the OSD would be cool!  :)

p.s. I've written the code for 3D throttle checking and currently testing.  The earlier lack of 3D support would only have the side affect of possibly deactivating the protection too early and not some other affect like erroneous triggering.  So no worse than before but possibly not better - but I'm fixing it now.

mikeller commented:
Sorry, I thought the current name was still `runaway_takeoff_prevention` like it used to be.

>  Also I don't think "freakout" clearly identifies the event at takeoff. Could just as easily describe a de-sync behavior to a new user.

The exactly same can also be said about `antitaz_`. Without prior knowledge I doubt that many new users will be able to make the link from `taz` to 'tasmanian devil', much less know what it refers to before they have seen it happen.

> In support of antitaz_ - everybody I've talked to or has read a post where the term "anti-taz" was used knew immediately that I was talking about the "Tasmanian Devil" on takeoff.

Not sure if this also applies to newbie pilots (which are a big part of the 'target audience' of users who don't know about bench testing / don't know how to bench test).

> Lastly having "ANTI-TAZ" show up in the OSD would be cool! :)

It will also leave abovementioned newbie pilots clueless about why their craft does not work how they think it should work - again, something informative would be more helpful.

mikeller commented:
> I'm also quite sure that everybody is going to end up calling it anti-taz (just like everybody refers to "turtle mode" and not "flip over after crash") so how does the user equate "runaway_takeoff_" (or whatever we call them) when they hear about "anti-taz"?

There is nothing wrong with users 'in the know' calling it by a different name, they are in the know by definition.
Fact is, I have not seen (new) users ask 'what does "flip over after crash" stand for?', whereas we still get it almost weekly that users ask 'what does "moron threshold" do?'. I think making stuff findable / understandable to users who don't know about it beforehand is important.

Saving flash at the cost of having to answer to users asking avoidable questions is not a good tradeoff in my opinion.

I don't know where to take this from here. I was hoping for somebody to come up with a name that is catchy **and** informative. In absence of this I'd still be in favour of `runaway_takeoff_prevention`. While it is not the perfect description, it still gives users with no prior exposure an idea of what it is for. But if majority opinion is that we should go for a non-descriptive name then I'll accept that.

etracer65 commented:
Ok, getting back to the code for now...  I need some help with the 3D stuff.  Plain old "normal" 3D mode is fine and the code is working.  However I need to understand the modes "3D ON A SWITCH" and "DISABLE 3D".

"DISABLE 3D" I think is working correctly in that with the mode off I can only arm within the 3D deadband (like full 3D mode).  When the mode is active then I have to be below min_check to arm (like non-3D mode).  This makes sense if I think I understand the purpose of the mode.  The blackbox log makes it really confusing.  The logging doesn't seem to understand the DISABLE 3D mode and still writes rcCommand throttle values apparently scaled to the 3D space from only 1500-2000us.

"3D ON A SWITCH" - I have no idea what this is trying to do.  When this mode is configured it's only possible to arm with the throttle below min_check regardless of whether the mode is active or not.  So I'm not sure if this is a bug or I'm just not understanding.

I based my code on the this in rc_controls.c

```
throttleStatus_e calculateThrottleStatus(void)
{
    if (feature(FEATURE_3D)) {
        if (IS_RC_MODE_ACTIVE(BOX3DDISABLE) || isModeActivationConditionPresent(BOX3DONASWITCH)) {
            if (rcData[THROTTLE] < rxConfig()->mincheck) {
                return THROTTLE_LOW;
            }
        } else if ((rcData[THROTTLE] > (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle) && rcData[THROTTLE] < (rxConfig()->midrc + flight3DConfig()->deadband3d_throttle))) {
            return THROTTLE_LOW;
        }
    } else if (rcData[THROTTLE] < rxConfig()->mincheck) {
        return THROTTLE_LOW;
    }

    return THROTTLE_HIGH;
}
```
It seems like the purpose of both modes is to disable 3D mode as the code will use the min_check comparison if either are active.  Not sure why there are 2 different modes to do this or how they differ.

I translated this into:

```
// calculate the throttle stick percent - integer math is good enough here.
uint8_t calculateThrottlePercent(void)
{
    int8_t ret = 0;
    if (feature(FEATURE_3D)
        && !IS_RC_MODE_ACTIVE(BOX3DDISABLE)
        && !isModeActivationConditionPresent(BOX3DONASWITCH)) {

        if ((rcData[THROTTLE] >= PWM_RANGE_MAX) || (rcData[THROTTLE] <= PWM_RANGE_MIN)) {
            ret = 100;
        } else {
            if (rcData[THROTTLE] > (rxConfig()->midrc + flight3DConfig()->deadband3d_throttle)) {
                ret = ((rcData[THROTTLE] - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle) * 100) / (PWM_RANGE_MAX - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle);
            } else if (rcData[THROTTLE] < (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle)) {
                ret = ((rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - rcData[THROTTLE]) * 100) / (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - PWM_RANGE_MIN);
            }
        }
    } else {
        ret = constrain(((rcData[THROTTLE] - rxConfig()->mincheck) * 100) / (PWM_RANGE_MAX - rxConfig()->mincheck), 0, 100);
    }
    return ret;
}
```

Which seems correct based on the reference code.  But until I better understand the purpose of these two modes I don't really trust the logic.

mikeller commented:
The modes are defined as follows:
- `DISABLE_3D`: Act like `FEATURE_3D` was not enabled;
- `3D_ON_A_SWITCH`: Use the switch to switch between normal and inverted flight, with the throttle input -> output mapping being the same as non-3D for both switch positions, but rotation direction being inverted when the switch is on.

In essence, I think if `DISABLE_3D` is active, or 3D_ON_A_SWITCH` is configured, your 'stick idle' detection should work the same way as if 3D was not enabled - and from that perspective your code looks fine.

I'd still like a 3D flyer to test and confirm that the deadband evaluation that you are doing actually works as it should.

etracer65 commented:
Ok, that makes sense.  In that case it looks like the code is working as expected.  I just have to do a could of flight tests to make sure normal mode still works as expected.  I can't test 3D mode so we'll have to try and recruit someone.

Once I test I'll commit the 3D code along with the default change for `runaway_takeoff_activate_delay` and the removal of the two unnecessary parameters:

`runaway_takeoff_deactivate_pidlimit` - hard-code to 10
`runaway_takeoff_deactivate_stick_percent `- hard-code to 15

etracer65 commented:
Even though I can't flight test 3D mode, I did write my calculated throttle percent to a debug value in blackbox to test the calculation.  Everything was working as it should.  The 3D throttle calculation is basically identical to the stick deflection calculation - looking for the absolute percentage of deflection from mid_rc.

fiveangle commented:
Perhaps @brycedjohnson would be willing to test the feature on an actual 3D quad ?

brycedjohnson commented:
@fiveangle I've hardly been able to fly lately - getting ready to move to CO and we have had a streak of cold weather where the highs are <0F in MN.

fiveangle commented:
@brycedjohnson - gotcha ! Feel free to send some of that cold here to California so it stops raining what little snow we have in Tahoe ;)  🏂 ❄️

etracer65 commented:
I've committed the changes to add 3D mode handling to the throttle percentage calculation, updated a default and removed the two unnecessary configuration parameters.  At this point I'm done and I feel like it's ready to merge.

@mikeller @martinbudden What do we need to do about the couple of targets that are still overflowing?  Is there anything I can change to reduce the footprint?  It looks like BETAFLIGHTF3 is overflowing by a lot - not sure why it's so bad.  OMNIBUS is also but not by so much.  Not sure if there any more since the Travis job terminates.

Is there a way to no have a local `make all` terminate if one of the targets fails to see if there are more?

mikeller commented:
@etracer65: @martinbudden said he had some ideas to reduce flash space usage, and there is hope that this will solve the shortage. If not, we'll have to have a look and find more features to cull on the problematic targets - but we'll do this in a separate pull request, since there are more pull requests stuck in the flash space shortage at the moment.

And no, there isn't. But you can use `make targets-group{12345}` and `make targets-group-rest` to build the targets split up the same way as CI does.

mikeller commented:
@etracer65: Can this be rebased / squashed? We've changed optimisation settings for F3, so this might fit when rebased against latest master.

etracer65 commented:
@mikeller Done.  Was able to successfully re-sync my fork from upstream and rebase/squash.  There were a couple of small conflicts with later commits, but no problem to fix.

Unfortunately OMNIBUS is still oversize by 981 bytes, but at least the others appear to be okay now.

I don't really understand why my code appears to use so much space.  On the surface there doesn't appear to enough code to use that much extra space.  As a test I commented out the main code in fc_core.c and then OMNIBUS was able to compile without a problem.  So maybe someone can look at my implementation in fc/fc_core.c and suggest ways to reduce the footprint.  I did have to include math.h to make use of fabsf() but removing that and switching to implicit double conversions and using abs() actually used about 10 more bytes so I don't think that's the problem.

etracer65 commented:
Throwing out a few ideas to reduce the code size in general...

I found that in the current 3.3 master changing fc/fc_core.c from speed optimized to size optimized resulted in 4776 bytes of savings for target OMNIBUS.  While there are parts in fc_core.c that should be speed optimized (like the `taskMainPidLoop()`, `subTaskPidController()` and `subTaskMotorUpdate()`, etc.) there are many other parts that don't need the speed optimization like `processRx()`.  So maybe two  different options:

1. Split out the functions that don't need speed optimization into a separate file and make them size optimized.
2. Move the speed-critical functions either to a separate file or possibly into flight/pid.c which is also speed optimized.

I realize changes like this shouldn't be part of this pull request, but just throwing out ideas to address the space issue for F3's in general.

etracer65 commented:
As a test I created https://github.com/betaflight/betaflight/pull/5061 for discussion.  It moves the time-critical tasks from fc_core.c to fc_tasks.c.  Then fc_core.c can be changed to size-optimized.  Saves 3288 bytes.

mikeller commented:
@etracer65: This needs to be rebased now.
@martinbudden: Is this ok now or are there more changes needed?

etracer65 commented:
@mikeller @martinbudden Tying to find more optimizations to fit this into OMNIBUS (assuming we don't retain https://github.com/betaflight/betaflight/pull/5061).

https://github.com/betaflight/betaflight/pull/5068

Saves 72 bytes in master now, will save roughly double that when incorporated into this PR as the code calculated the pidSum in two additional places.

martinbudden commented:
@etracer65 , now that you have got this working you could save significant ROM by:
1. Removing most or all of the DEBUG code (ie the `DEBUG_SET` statements.
2. Remove the most or all of the configuration settings and hardcode the values.

(2), assumes you are reasonably confident that you have found good values for the config values.

etracer65 commented:
@martinbudden I'm not really comfortable with either at this point.  There have only been a handful of pilots flying this and while we've encountered no problems, I don't think it's a big enough sample set.  My plan was to finalize this sometime during the RC stages.

The DEBUG is critical to understanding what happened if the code behaves unexpectedly so I wouldn't want to remove that yet.  The debug values give visibility into the state machine so we can can compare with the pidSum and timing.

I have a few other optimizations (along with the pidSum calc optimize https://github.com/betaflight/betaflight/pull/5068) that will get us down to around 600 bytes overflow (from 981 now).

I need to rebase so I'm stuck on the code until we decide whether https://github.com/betaflight/betaflight/pull/5061 is reverted or not.

martinbudden commented:
Removing them in the RC phase sounds fine.

mikeller commented:
@martinbudden: Can you approve your review then?

etracer65 commented:
@mikeller @martinbudden I rebased and then made some optimizations.  Got the overflow for OMNIBUS reduced from 981 to 629 bytes. Next I'll work on adding `USE_RUNAWAY_TAKEOFF` conditional compilation flag.  From the conversation it looks like for now I'll leave it enabled in OMNIBUS and we'll look for other flash savings.

I'll the commits separate for easier review and than squash at the end.

mikeller commented:
@etracer65: Good work! I'd try to remove features from OMNIBUS, best in a separate pull request. My choices would be (in reverse order of importance):

```
USE_RX_MSP
USE_RTC_TIME
USE_COPY_PROFILE_CMS_MENU
USE_RTC_TIME
USE_HUFFMAN
USE_TELEMETRY_HOTT
USE_CAMERA_CONTROL
```

etracer65 commented:
Any suggestions on where in `target/common_fc_pre.h` the `#define USE_RUNAWAY_TAKEOFF` should go?  Doesn't really seem like it needs to be limited by processor type.  Also flash size doesn't seem necessary as surprisingly the code compiles and doesn't overflow on target NAZE (which I realize isn't officially supported anymore).  I'm not sure the define needs to be inside of any conditions.

mikeller commented:
@etracer65: Into `#if (FLASH_SIZE > 64)` in this case.

etracer65 commented:
@mikeller @martinbudden In the latest commit I've added conditional compilation support with `USE_RUNAWAY_TAKEOFF`.  I also fixed the unit test `arming_prevention_unittest` that was failing because it needed `axisPIDSum[]` stubbed in.

I wrote some scripting to test compiling all the supported targets locally (logs fails and continues).  With `USE_RUNAWAY_TAKEOFF` defined I verified that every target compiles except OMNIBUS which is still 629 bytes over (as expected).  With it undefined every target compiles successfully.

Once review is complete I'll squash the commits.

etracer65 commented:
@mikeller 

> USE_RX_MSP
USE_RTC_TIME
USE_COPY_PROFILE_CMS_MENU
USE_RTC_TIME
USE_HUFFMAN
USE_TELEMETRY_HOTT
USE_CAMERA_CONTROL

`USE_RX_MSP` only frees up 152 bytes.
`USE_RTC_TIME` frees up a bunch and reduces OMNIBUS to:

```
Linking OMNIBUS 
   text	   data	    bss	    dec	    hex	filename
 248261	   9024	  31432	 288717	  467cd	./obj/main/betaflight_OMNIBUS.elf
Creating HEX ./obj/betaflight_3.3.0_OMNIBUS.hex 
```

And this is with Runaway Takeoff included!  Do we want to remove `USE_RTC_TIME` from OMNIBUS in a separate PR?  Also the bulk of the space used is in common/time.c and there may be opportunities for optimization.

mikeller commented:
@etracer65: Yes, please open a separate pull request for the feature removal. `USE_RTC_TIME` is a 'nice to have' feature, and not having it will not affect flying related performance or capabilities.

Re optimising `common/time.c`, this was taken 1:1 from iNav, so optimisation should probably be done in a way that makes it possible to keep these files in synch...

etracer65 commented:
@mikeller https://github.com/betaflight/betaflight/pull/5075 created.

After it gets merged let me know when the review of the recent commits to this PR is complete and I'll rebase/squash.  After that the travis builds should succeed.

mikeller commented:
@etracer65: Done.

mikeller commented on src/main/fc/fc_core.c:
```diff
@@ -389,6 +424,55 @@ static bool canUpdateVTX(void)
+}
+#endif
+
+#ifdef USE_RUNAWAY_TAKEOFF
```
These two `#ifdef` sections could be combined.

mikeller commented on src/main/fc/runtime_config.c:
```diff
@@ -32,7 +32,11 @@ static uint32_t enabledSensors = 0;
     "THROTTLE", "ANGLE", "BOOTGRACE", "NOPREARM", "LOAD",
-    "CALIB", "CLI", "CMS", "OSD", "BST", "MSP", "ARMSWITCH"
+    "CALIB", "CLI", "CMS", "OSD", "BST", "MSP",
+#ifdef USE_RUNAWAY_TAKEOFF
```
This should not be conditional. If it is, the numbering of flags after it will become target dependent, creating problems for interfacing apps and documentation.

mikeller commented on src/main/fc/runtime_config.h:
```diff
@@ -51,10 +51,20 @@ typedef enum {
     ARMING_DISABLED_OSD_MENU        = (1 << 13),
     ARMING_DISABLED_BST             = (1 << 14),
     ARMING_DISABLED_MSP             = (1 << 15),
+#ifdef USE_RUNAWAY_TAKEOFF
```
This should not be conditional. If it is, the numbering of flags after it will become target dependent, creating problems for interfacing apps and documentation.

martinbudden replied:
Agreed. I think all `USE_RUNAWAY_TAKEOFF` instances should be removed from this header file.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -113,6 +113,13 @@ PG_DECLARE_ARRAY(pidProfile_t, MAX_PROFILE_COUNT, pidProfiles);
 
 typedef struct pidConfig_s {
     uint8_t pid_process_denom;              // Processing denominator for PID controller vs gyro sampling rate
+#ifdef USE_RUNAWAY_TAKEOFF
```
This cannot be conditional. The goal is to use parameter groups to programmatically generate interfaces (to be used over MSP). This means that every version of a parameter group has to be identical for all targets that support it.

martinbudden commented:
@etracer65 , I've approved this, since you've addressed my comments. However, there are still some outstanding comments by @mikeller that need to be addressed.

etracer65 commented:
@mikeller I've made the requested changes to the conditional compilation.  If these are good let me know and I'll rebase/squash.

mikeller commented:
@etracer65: Yes, looks good to me now.

etracer65 commented:
@mikeller Successfully rebased/squashed.  Just waiting for travis to do it's thing.  OMNIBUS compiles locally now without any problems.

etracer65 commented:
@mikeller Just an interesting observation as I've been trying to optimize space...

Previously you commented to me that variable initialization to zero in the declaration wasn't necessary but would get optimized away by the compiler anyway (don't remember if this was on slack or here).  For example the code:

```
uint8_t calculateThrottlePercent(void)
{
    uint8_t ret = 0;
    if (feature(FEATURE_3D)
        && !IS_RC_MODE_ACTIVE(BOX3DDISABLE)
        && !isModeActivationConditionPresent(BOX3DONASWITCH)) {

        if ((rcData[THROTTLE] >= PWM_RANGE_MAX) || (rcData[THROTTLE] <= PWM_RANGE_MIN)) {
            ret = 100;
        } else {
            if (rcData[THROTTLE] > (rxConfig()->midrc + flight3DConfig()->deadband3d_throttle)) {
                ret = ((rcData[THROTTLE] - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle) * 100) / (PWM_RANGE_MAX - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle);
            } else if (rcData[THROTTLE] < (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle)) {
                ret = ((rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - rcData[THROTTLE]) * 100) / (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - PWM_RANGE_MIN);
            }
        }
    } else {
        ret = constrain(((rcData[THROTTLE] - rxConfig()->mincheck) * 100) / (PWM_RANGE_MAX - rxConfig()->mincheck), 0, 100);
    }
    return ret;
}
```

As is with `uint8_t ret = 0;` it compiles to:

```
   text	   data	    bss	    dec	    hex	filename
 248281	   9024	  31528	 288833	  46841	./obj/main/betaflight_OMNIBUS.elf
```

Since the initialization to zero is not needed in this code because every condition assigns a value, the code could be changed to `uint8_t ret;` as:

```
uint8_t calculateThrottlePercent(void)
{
    uint8_t ret;
    if (feature(FEATURE_3D)
        && !IS_RC_MODE_ACTIVE(BOX3DDISABLE)
        && !isModeActivationConditionPresent(BOX3DONASWITCH)) {

        if ((rcData[THROTTLE] >= PWM_RANGE_MAX) || (rcData[THROTTLE] <= PWM_RANGE_MIN)) {
            ret = 100;
        } else {
            if (rcData[THROTTLE] > (rxConfig()->midrc + flight3DConfig()->deadband3d_throttle)) {
                ret = ((rcData[THROTTLE] - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle) * 100) / (PWM_RANGE_MAX - rxConfig()->midrc - flight3DConfig()->deadband3d_throttle);
            } else if (rcData[THROTTLE] < (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle)) {
                ret = ((rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - rcData[THROTTLE]) * 100) / (rxConfig()->midrc - flight3DConfig()->deadband3d_throttle - PWM_RANGE_MIN);
            }
        }
    } else {
        ret = constrain(((rcData[THROTTLE] - rxConfig()->mincheck) * 100) / (PWM_RANGE_MAX - rxConfig()->mincheck), 0, 100);
    }
    return ret;
}
```

With no other changes it compiles to:

```
   text	   data	    bss	    dec	    hex	filename
 248297	   9024	  31528	 288849	  46851	./obj/main/betaflight_OMNIBUS.elf
```

Actually uses 16 more bytes!  Not really sure why this is - maybe because this code is compiled with speed optimization.  But it just goes to show that sometimes the compiler behaves in mystical ways...

mikeller commented:
@etracer65: Wow, that's an interesting observation. I wonder if this has to do with range checking (one would expect there to be no difference between the two versions), and will go away if `ret` is changed to `int`?

etracer65 commented:
@mikeller Just tested and changing to `int` causes it to be the larger size regardless of whether it's initialized or not.

martinbudden commented:
Two points:

1. You need to look at the assembly output of the compiler to see what is going on.
2. The code increase is probably less than 16 bytes - code is "blocked" into units on (I think) 16 byte boundaries.

etracer65 commented:
I guess the point is that it's counter-intuitive regardless of the actual number of bytes.  The only way I found this was trying to reduce flash usage by thinking that removing the unnecessary initialization might save a few bytes (or at least have no effect).  I was surprised when it increased flash size.  So the lesson is that looking at code like this one might be tempted to remove the unnecessary initialization in the name of optimization and actually be accomplishing the reverse.

etracer65 commented:
And YAY!  The travis run completed and every target finally compiled!  I think we're good to go.

martinbudden commented:
A general rule about optimisation (for speed or size) is that it can be counter-intuitive so that it is important to check and/or test your results. Thing you think will reduce ROM size can increase it, and things that you think will speed things up can slow them down. Testing and checking is essential.

Another point is that it is not worth making the code less understandable or more fragile just to save a few bytes of ROM.

mikeller commented:
:tada:

etracer65 commented:
@martinbudden I originally posted this under https://github.com/betaflight/betaflight/pull/5092 and you might have missed it.  Following up to make sure there are no issues with changes to parameter groups made here.

---
So understand for the future, what circumstances require changing the PG version?  Is it only if an existing parameter changes in data type?  What about if additional parameters are added or removed?

I'm concerned because I didn't do anything with the version when adding the parameters for Runaway Takeoff.  In that case parameters were added to flight/pid.c

```
PG_REGISTER_WITH_RESET_TEMPLATE(pidConfig_t, pidConfig, PG_PID_CONFIG, 1);

#ifdef STM32F10X
#define PID_PROCESS_DENOM_DEFAULT       1
#elif defined(USE_GYRO_SPI_MPU6000) || defined(USE_GYRO_SPI_MPU6500)  || defined(USE_GYRO_SPI_ICM20689)
#define PID_PROCESS_DENOM_DEFAULT       4
#else
#define PID_PROCESS_DENOM_DEFAULT       2
#endif

#ifdef USE_RUNAWAY_TAKEOFF
PG_RESET_TEMPLATE(pidConfig_t, pidConfig,
    .pid_process_denom = PID_PROCESS_DENOM_DEFAULT,
    .runaway_takeoff_prevention = true,
    .runaway_takeoff_threshold = 60,            // corresponds to a pidSum value of 60% (raw 600) in the blackbox viewer
    .runaway_takeoff_activate_delay = 75,       // 75ms delay before activation (pidSum above threshold time)
    .runaway_takeoff_deactivate_throttle = 25,  // throttle level % needed to accumulate deactivation time
    .runaway_takeoff_deactivate_delay = 500     // Accumulated time (in milliseconds) before deactivation in successful takeoff
);
#else
PG_RESET_TEMPLATE(pidConfig_t, pidConfig,
    .pid_process_denom = PID_PROCESS_DENOM_DEFAULT
);
#endif
```

Also complicated by the conditional compilation changing the parameter content.

martinbudden commented:
@etracer65 , the PG version needs to change if the offset of any of the parameters changes - that is if parameters in the middle are added, removed or change in side.

If you just add parameters on to the end of a PG, then the version doesn't need to change.

etracer65 commented:
Okay, makes sense.  Thanks.

mikeller commented:
Probably as a (soft) add on to @martinbudden's statement: If the way a parameter is used is changed (e.g. different scaling, but the same type), the version _should_ probably also be changed, to keep legacy clients from storing bogus data.

etracer65 commented:
As usual, @ozzy33 managed to have extreme flight characteristics and caused false triggering of Runaway Takeoff Prevention. Trouble maker! :)

Actually nothing was wrong with the code.  The deactivation never happened because the throttle never exceeded 25% (generally stayed below 20%!).  Secondarily airmode never activated because the throttle stayed so low.  This led to the pidSum growing rapidly on a hard yaw move (at very low throttle) because without airmode active the PID controller was lazy.  Since the Runaway Takeoff Prevention didn't deactivate it triggered on the high yaw pidSum.

So two takeaways:

1. In this case with a really light and/or high powered quad the `runaway_takeoff_deactivate_throttle_percent` needs to be lower - probably around 15% (!)
2. Airmode needs to get activated - probably change `airmode_throttle_start_percent` to a lower value.

For those interested here's the log. 

[BTFL_BLACKBOX_LOG_falls out of the sky.BBL.zip](https://github.com/betaflight/betaflight/files/1709360/BTFL_BLACKBOX_LOG_falls.out.of.the.sky.BBL.zip)

Ozzy33 commented:
I do seem to find the edge cases do to my style of LOS flying :)
I am not a fan of anti-taz name, only because I invented the Tasmanian Devil (Taz for short) trick 9 months ago. And variations, the Taz-Mowian Devil, mowing the grass with the bottom prop (in my LOS trick compilation vid) and Spaz-manian Devil :)
https://www.youtube.com/watch?v=XBmVvttGlCU&t=71s

fiveangle commented:
> I am not a fan of anti-taz name, only because I invented the Tasmanian Devil

Even more proof that most people fully associate the "Taz" cartoon reference with uncontrollable spaz-out motion :)

Ozzy33 commented:
The other taz was when BF introduced airmode (its fixed now), if you clipped the ground the quad would sometimes just freak out all over the place, just like taz, lol
